pkg/pillar/cmd/zedagent/handlevolume.go

```diff
@@ -128,3 +131,47 @@ func handleVolumeStatusDelete(ctxArg interface{},
 	PublishVolumeToZedCloud(ctx, uuidStr, nil, ctx.iteration)
 	ctx.iteration++
 }
+
+func createVolumeInstanceMetrics(ctx *getconfigContext, reportMetrics *metrics.ZMetricMsg) {
+	log.Debugf("Volume instance metrics started")
+	sub := ctx.subVolumeStatus
+	volumelist := sub.GetAll()
+	if volumelist == nil || len(volumelist) == 0 {
+		return
+	}
+	for _, volume := range volumelist {
+		volumeStatus := volume.(types.VolumeStatus)
+		volumeMetric := new(metrics.ZMetricVolume)
+		volumeMetric.Uuid = volumeStatus.VolumeID.String()
+		volumeMetric.DisplayName = volumeStatus.DisplayName
+		getVolumeResourcesMetrics(volumeStatus, volumeMetric)
+		reportMetrics.Vm = append(reportMetrics.Vm, volumeMetric)
+	}
+	log.Debugf("Volume instance metrics done: %v", reportMetrics.Vm)
+}
+
+func getVolumeResourcesMetrics(volumeStatus types.VolumeStatus,
+	volumeMetric *metrics.ZMetricVolume) error {
+
+	if volumeStatus.IsContainer() {
+		// XXX For container volumes, max size is coming zero
+		// from the controller. So for now, we are setting up
+		// max size and the current size equal to the downloaded size
+		size, err := utils.DirSize(volumeStatus.FileLocation)
+		if err != nil {
+			return err
+		}
+		volumeMetric.TotalBytes = size
+		volumeMetric.UsedBytes = size
+		volumeMetric.FreeBytes = size
+	} else {
+		imgInfo, err := diskmetrics.GetImgInfo(volumeStatus.FileLocation)
```

Comment:
This we have to call periodically, so perhaps we should do that periodic DirSize for the container as well since it can grow? Thoughts?

Reply:
We are already calculating the directory size for the container every time. Am I missing something?

Reply:
Conceptually we have a maxsize for a container; we just don't happen to enforce it. (And it might be set to the same as the actual size if the controller left the maxsize at zero.) But if the controller set maxsize = N, shouldn't we report that as the max, and report free as max-actual? 
We already have a non-zero maxsize in the volumestatus, right?

Reply:
Let's worry about that later.